plugin/output/s3/s3.go

```diff
@@ -97,96 +120,97 @@ func (p *Plugin) Start(config pipeline.AnyConfig, params *pipeline.OutputPluginP
 		longpanic.Go(p.uploadWork)
 		longpanic.Go(p.compressWork)
 	}
 
-	// initialize minio client object.
-	minioClient, err := minio.New(p.config.Endpoint, p.config.AccessKey, p.config.SecretKey, p.config.Secure)
-	if err != nil || minioClient == nil {
-		p.logger.Panicf("could not create minio client, error: %s", err.Error())
-	}
-	p.client = minioClient
-
-	if p.config.client != nil {
-		p.logger.Info("set mock client")
-		p.client = *p.config.client
-	}
-
-	exist, err := p.client.BucketExists(p.config.Bucket)
-	if err != nil {
-		p.logger.Panicf("could not check bucket: %s, error: %s", p.config.Bucket, err.Error())
-	}
-	if !exist {
-		p.logger.Fatalf("bucket: %s, does not exist", p.config.Bucket)
-	}
-	p.logger.Info("client is ready")
-	p.logger.Infof("bucket: %s exists", p.config.Bucket)
-
-	anyPlugin, _ := file.Factory()
-	p.outPlugin = anyPlugin.(*file.Plugin)
+	// initialize minio clients.
+	p.clients = p.createMinioClients()
+	p.startPlugins(params, outPlugCount)
 
-	p.outPlugin.SealUpCallback = p.addFileJob
-
-	p.outPlugin.Start(&p.config.FileConfig, params)
 	p.uploadExistingFiles()
 }
 
 func (p *Plugin) Stop() {
-	p.outPlugin.Stop()
+	p.outPlugins.Stop()
 }
 
 func (p *Plugin) Out(event *pipeline.Event) {
-	p.outPlugin.Out(event)
+	p.outPlugins.Out(event, pipeline.PlugSelector{
+		CondType:  pipeline.ByNameSelector,
+		CondValue: p.getS3BucketFromEvent(event),
+	})
 }
 
-// uploadExistingFiles gets files from dirs, sorts it, compresses it if it's need, and then upload to s3
-func (p *Plugin) uploadExistingFiles() {
-	// get all compressed files
-	pattern := fmt.Sprintf("%s*%s", p.targetDir, p.compressor.getExtension())
-	compressedFiles, err := filepath.Glob(pattern)
-	if err != nil {
-		p.logger.Panicf("could not read dir: %s", p.targetDir)
-	}
-	// sort compressed files by creation time
-	sort.Slice(compressedFiles, p.getSortFunc(compressedFiles))
-	// upload archive
-	for _, z := range compressedFiles {
-		p.uploadCh <- z
+// getS3BucketFromEvent tries finds bucket in multi_buckets. Otherwise, events goes to main bucket.
+func (p *Plugin) getS3BucketFromEvent(event *pipeline.Event) string {
+	bucket := event.Root.Dig(bucketField).AsString()
+	if !p.outPlugins.Exists(bucket) {
+		return p.config.Bucket
 	}
+	return bucket
+}
 
-	// compress all files that we have in the dir
-	p.compressFilesInDir()
+// uploadExistingFiles gets files from dirs, sorts it, compresses it if it's need, and then upload to s3.
+func (p *Plugin) uploadExistingFiles() {
+	for bucketName, dir := range p.targetDirs {
+		// get all compressed files.
+		pattern := fmt.Sprintf("%s*%s", dir, p.compressor.getExtension())
+		compressedFiles, err := filepath.Glob(pattern)
+		if err != nil {
+			p.logger.Panicf("could not read dir: %s", dir)
+		}
+		// sort compressed files by creation time.
+		sort.Slice(compressedFiles, p.getSortFunc(compressedFiles))
+		// upload archive.
+		for _, z := range compressedFiles {
+			splitPath := strings.Split(z, "/")
```

Comment:
It seems more readable to use filepath package.
```
dir := filepath.Dir(p)
parent := filepath.Base(dir)
```

Also here need to process case of unexpected dirs structure.

Reply:
Good point